src/rust/engine/ui/src/display.rs

```diff
@@ -0,0 +1,307 @@
+      // as we've done here is the safest way to avoid terminal oddness.
+      cursor_start: (1, 1),
+      terminal_size: EngineDisplay::get_size(),
+    }
+  }
+
+  // Gets the current terminal's cursor position, if applicable.
+  fn get_cursor_pos(&mut self) -> (u16, u16) {
+    match self.terminal {
+      Console::Terminal(ref mut t) => match &t.cursor_pos() {
+        Ok((x, y)) => (*x, *y),
+        // N.B. Real TTY coordinates start at 1.
+        Err(_) => (0, 0),
+      },
+      Console::Pipe(_) => (0, 0),
+    }
+  }
+
+  // Gets the current terminal's width and height, if applicable.
+  fn get_size() -> (u16, u16) {
+    match termion::terminal_size() {
+      Ok((x, y)) => (x, y),
+      Err(_) => (0, 0),
+    }
+  }
+
+  // Whether or not the EngineDisplay is running (whether .start() has been called).
+  pub fn is_running(&self) -> bool {
+    self.running
+  }
+
+  // Sets the terminal size per-render for signal-free resize detection.
+  fn set_size(&mut self) {
+    self.terminal_size = EngineDisplay::get_size();
+  }
+
+  // Sets the terminal size for signal-free resize detection.
+  fn get_max_log_rows(&self) -> usize {
+    self.terminal_size.1 as usize - self.action_map.len() - 1
+  }
+
+  // Prep the screen for painting by clearing it from the cursor start position.
+  fn clear(&mut self) {
+    let cursor_start = self.cursor_start;
+    self
+      .write(format!(
+        "{goto_origin}{clear}",
+        goto_origin = cursor::Goto(cursor_start.0, cursor_start.1),
+        clear = clear::AfterCursor,
+      ))
+      .expect("could not write to terminal");
+  }
+
+  // Flush terminal output.
+  fn flush(&mut self) -> Result<()> {
+    match self.terminal {
+      Console::Terminal(ref mut t) => t.flush(),
+      Console::Pipe(ref mut p) => p.flush(),
+    }
+  }
+
+  // Writes output to the terminal.
+  fn write(&mut self, msg: String) -> Result<usize> {
+    match self.terminal {
+      Console::Terminal(ref mut t) => t.write(msg.as_bytes()),
+      Console::Pipe(ref mut p) => p.write(msg.as_bytes()),
+    }
+  }
+
+  // Renders a divider between the logs and action output.
+  fn render_divider(&mut self, offset: u16) {
+    let cursor_start = self.cursor_start;
+    let padding = self.padding.clone();
+    let divider = self.divider.clone();
+
+    self
+      .write(format!(
+        "{pos}{clear_line}{padding}{blue}{divider}{reset}",
+        pos = cursor::Goto(1, cursor_start.1 + offset),
+        clear_line = clear::CurrentLine,
+        padding = padding,
+        blue = color::Fg(color::Blue),
+        divider = divider,
+        reset = color::Fg(color::Reset)
+      ))
+      .expect("could not write to terminal");
+  }
+
+  // Renders one frame of the log portion of the screen.
+  fn render_logs(&mut self, max_entries: usize) -> usize {
+    let cursor_start = self.cursor_start;
+    let printable_logs: Vec<String> = self.logs.clone().into_iter().take(max_entries).collect();
```

Comment:
it'll always contain enough logs to be able to fill on re-painting if the screen is resized larger - currently unbounded (but I'll cap this arbitrarily to 500 for now).

when I attempt to `s/.clone().into_iter()/.iter()/` I get:

```
error[E0502]: cannot borrow `*self` as mutable because `self.logs` is also borrowed as immutable
   --> ui/src/display.rs:168:7
    |
155 |     let printable_logs: Vec<&String> = self.logs.iter().take(max_entries).collect();
    |                                        --------- immutable borrow occurs here
...
168 |       self
    |       ^^^^ mutable borrow occurs here
...
183 |   }
    |   - immutable borrow ends here

error[E0502]: cannot borrow `*self` as mutable because `self.logs` is also borrowed as immutable
   --> ui/src/display.rs:179:7
    |
155 |     let printable_logs: Vec<&String> = self.logs.iter().take(max_entries).collect();
    |                                        --------- immutable borrow occurs here
...
179 |       self.render_divider(counter as u16);
    |       ^^^^ mutable borrow occurs here
...
183 |   }
    |   - immutable borrow ends here

error: aborting due to 2 previous errors
```

and am unable to satisfy the borrow checker via standard tricks.